docs/configuration.md

```diff
@@ -0,0 +1,153 @@
+    ["sketchbook" directory][sketchbook directory]. Library Manager
+    installations are made to the `libraries` subdirectory of the user
+    directory.
+* `logging` - configuration options for Arduino CLI's logs.
+    * `file` - path to the file where logs will be written.
+    * `format` - output format for the logs. Allowed values are `text` or
+    `json`.
+    * `level` - messages with this level and above will be logged. Valid levels
+    are: `trace`, `debug`, `info`, `warn`, `error`, `fatal`, `panic`.
+* `telemetry` - settings related to the collection of data used for continued
+improvement of Arduino CLI.
+    * `addr` - TCP port used for telemetry communication.
+    * `enabled` - controls the use of telemetry.
+
+## Configuration methods
+
+Arduino CLI may be configured in three ways:
+
+1. Command line flags
+1. Environment variables
+1. Configuration file
+
+If a configuration option is configured by multiple methods, the value set by
+the method highest on the above list overwrites the ones below it.
+
+If a configuration option is not set, Arduino CLI uses a default value.
+
+[`arduino-cli config dump`][arduino-cli config dump] displays the current
+configuration values.
+
+### Command line flags
+
+Arduino CLI's command line flags are documented in the command line help and the
+[Arduino CLI command reference].
+
+#### Example
+
+Setting an additional Board Manager URL using the
+[`--additional-urls`][arduino-cli global flags] command line flag:
+
+```shell
+$ arduino-cli core update-index --additional-urls https://downloads.arduino.cc/packages/package_staging_index.json
+```
+
+### Environment variables
+
+All configuration options can be set via environment variables. The variable
+names start with `ARDUINO`, followed by the configuration key names, with each
+component separated by `_`. For example, the `ARDUINO_DIRECTORIES_USER`
+environment variable sets the `directories.user` configuration option.
+
+On Linux or macOS, you can use the [`export` command][export command] to set
+environment variables. On Windows cmd, you can use the
+[`set` command][set command].
+
+#### Example
+
+Setting an additional Board Manager URL using the
+`ARDUINO_BOARD_MANAGER_ADDITIONAL_URLS` environment variable:
+
+```sh
+$ export ARDUINO_BOARD_MANAGER_ADDITIONAL_URLS=https://downloads.arduino.cc/packages/package_staging_index.json
+```
+
+### Configuration file
+
+[`arduino-cli config init`][arduino-cli config init] creates or updates a
+configuration file with the current configuration settings.
+
+This allows saving the options set by command line flags or environment
+variables. For example:
+
+```sh
+arduino-cli config init --additional-urls https://downloads.arduino.cc/packages/package_staging_index.json
+```
+
+#### File name
+
+The default configuration file name is `arduino-cli.yaml`. If you prefer another
+file name, use the [`--config-file` option][arduino-cli global flags].
+
+#### Supported formats
+
+* [YAML]
```

Comment:
Actually, we support all these formats:
JSON, TOML, YAML, HCL, envfile and Java properties config files
thanks to the spf13/Viper module!

Reply:
I'm very glad you raised this point. I was aware that Viper claimed to support all these, but when I tried to test it, I was disappointed to find I couldn't get any of the alternative formats other than JSON to be recognized. It turns out the problem was either a bug in Arduino CLI, or me not understanding the way it works. If there are multiple configuration files in the path, there is an order of precedence, even if you specify a file using `--config-file`:
```
$ ./arduino-cli config dump --verbose --config-file arduino-cli.toml
INFO[0000] Using config file: /home/per/ArduinoIDE/arduino-cli_nightly/arduino-cli.json 
```
(note that I specified the .toml file, but the .json file was used instead)
The problem was that I had a JSON file in the same folder as all the configuration files of other formats, so that JSON file always took precedence, even though the rest of the files were perfectly valid and supported.

I have updated the "Supported formats" section of the document accordingly.

Is this a bug? If so, I can submit a dedicated issue for it. If not, maybe we need to document this behavior? I'll concede it's likely a very rare occurrence for someone to have multiple configuration files. I did put the list of supported formats in their order of precedence.

The `--config-file` option's documentation seems somewhat misleading to me:
```
--config-file string        The custom config file (if not specified the default will be used).
```
From reading that, I would expect to be able to specify any filename, but it turns out that it's required to be `arduino-cli.whatever`:
```
$ ./arduino-cli config dump --verbose --config-file foo.json
INFO[0000] Config file not found, using default values 
```
Although you can specify the filename, with the current behavior it seems to be more of a config path option rather than a config file option:
```
$ ./arduino-cli config dump --verbose --config-file ./config
INFO[0000] Using config file: /home/per/ArduinoIDE/arduino-cli_nightly/config/arduino-cli.json
```
Specifying the filename only has an effect if the file doesn't exist:
```
$ ./arduino-cli config dump --verbose --config-file nonexistent.json
INFO[0000] Config file not found, using default values  
```
Otherwise, Arduino CLI disregards the filename and uses its preferred configuration file from that path.

I have updated the "File name" section of the document accordingly.

Reply:
Thanks @per1234  for pointing that out!
IMHO from the docs perspective, this PR is correct.
Regarding the config file loading, I think you discovered a bug :smile_cat: 
Please open an issue, as you said, whit this comment content. This is something we need to fix together wit other config file loading related small issues.